packages/langium/src/lsp/langium-grammar-hover-provider.ts

```diff
@@ -0,0 +1,83 @@
+import { AstNode } from '../syntax-tree';
+import { MaybePromise } from '../utils/promise-util';
+import { MultilineCommentHoverProvider } from './hover-provider';
+
+export class LangiumGrammarHoverProvider extends MultilineCommentHoverProvider {
+
+    readonly grammar: Grammar;
+    readonly documents: LangiumDocuments;
+    constructor(services: LangiumServices) {
+        super(services);
+        this.grammar = services.Grammar;
+        this.documents = services.shared.workspace.LangiumDocuments;
+    }
+
+    protected getAstNodeHoverContent(node: AstNode): MaybePromise<Hover | undefined> {
+        if(isCrossReference(node)) {
+            return this.getCrossReferenceHoverContent(node);
+        }
+        return super.getAstNodeHoverContent(node);
+    }
+
+    getHoverContent(document: LangiumDocument<AstNode>, params: HoverParams): MaybePromise<Hover | undefined> {
+        const rootNode = document.parseResult?.value?.$cstNode;
+        if (rootNode)  {
+            const offset = document.textDocument.offsetAt(params.position);
+            const cstNode = findLeafNodeAtOffset(rootNode, offset);
+            if (cstNode && cstNode.offset + cstNode.length > offset) {
+                const relevantNode = findRelevantNode(cstNode);
+                const targetNode = this.references.findDeclaration(cstNode);
+                if (isCrossReference(relevantNode)) {
+                    return this.getAstNodeHoverContent(relevantNode);
+                } else if (targetNode) {
+                    return this.getAstNodeHoverContent(targetNode?.element);
+                }
+            }
+        }
+        return undefined;
+    }
+
+    getCrossReferenceHoverContent(node: CrossReference): MaybePromise<Hover | undefined> {
+        const ref = node.type.ref;
+        if (ref) {
+            if (isType(ref)) {
+                return {
+                    contents: {
+                        kind: 'markdown',
+                        value: ref.$cstNode!.text
+                    }
+                };
+            } else if (isParserRule(ref)) {
+                const typeName = ref.name;
+                const terminal = node.terminal;
+                let terminalName = '';
+                if (terminal === undefined){
+                    const terminalRule = findNameAssignment(ref);
+                    if (terminalRule && isRuleCall(terminalRule.terminal)){
+                        terminalName = terminalRule.terminal.rule.ref!.name;
+                    }
+                } else if (isKeyword(terminal)) {
+                    terminalName = terminal.value;
+                } else if (isRuleCall(terminal)) {
+                    terminalName = terminal.rule.ref!.name;
+                }
+
+                return {
+                    contents: {
+                        kind: 'markdown',
+                        value: `[${typeName}:${terminalName}]`
```

Comment:
Suggestion: We should put this in a code block so that it receives correct syntax highlighting
```suggestion
                        value: '```\n' + `[${typeName}:${terminalName}]` + '\n```'
```